staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -647,10 +647,10 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
 		structuralSchemas[v.Name] = s
 	}
 
-	openAPIModels, err := buildOpenAPIModelsForApply(r.staticOpenAPISpec, crd)
+	typeConverter, err := buildTypeConverterForApply(crd.Spec.Versions, structuralSchemas, r.staticOpenAPISpec, crd.Spec.PreserveUnknownFields)
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("error building openapi models for %s: %v", crd.Name, err))
-		openAPIModels = nil
+		utilruntime.HandleError(fmt.Errorf("failed to building typeConverter for apply: %s: %v", crd.Name, err))
+		return nil, fmt.Errorf("the server could not properly serve the CR schema") // validation should avoid this
```

Comment:
This is the error we return in many other cases in this file, and then we log a more specific reason in the apiserver logs.

Not that we couldn't also return the specific error to the client

Reply:
That seems like something people can take an action on? "It failed because of XYZ", at least they can figure out what to do, no? Agreed that if it's a bug that requires an update/bug-fix then it's not as useful.

Reply:
If the validation of the crd's structural schema is done correctly, then this shouldn't ever be hit, so if it does get hit, then this would be due to a bug that requires an update/bug-fix

Reply:
This isn't meant to catch structural schema errors, it's meant to catch errors where, the structural schema is valid and it still fails to convert

Reply:
what does this do with non-structural schemas created via v1beta1? I'm concerned about turning this into a fatal error for the custom resource

Reply:
Non-structural schemas are handled as deduced schemas (+ apiversion, kind, and metadata)

Reply:
Because if the schema for a version fails structural validation, we just pass in an empty structural schema to the builder.AddStructural function

Reply:
https://github.com/kubernetes/kubernetes/blob/bb677ae6f655702fb7654f3b358a677d637d6514/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L1257-L1263

Reply:
rather than block all get/put/post/delete requests to the custom resource on this, I'd rather see us attempt this same operation in the controller and add an error condition to the CRD if it fails, to surface the error to the CRD author (similar to what we do for structural errors)

Reply:
I'm not against doing that for safety, but we don't expect this error to be able to be possibly triggered by users, since we ensure that only valid structural schemas are passed into the builder.AddStructural, which should only return an error if an invalid structural schema was passed in.

Reply:
If a user sends the apiserver a CRD with a schema that fails structural validation, then we won't even try to pass it into the builder.AddStructural function (and we pass in an empty structural schema instead).

Reply:
the surface area is large enough and the conversion complex enough that I think we should gain confidence with a non-blocking condition-adding approach before making it blocking